microprofile.ls/com.redhat.microprofile.ls/src/main/java/com/redhat/microprofile/ls/MicroProfileProjectInfoCache.java

```diff
@@ -187,32 +187,42 @@ public MicroProfileProjectInfoCache(MicroProfileProjectInfoProvider provider) {
 	 * @return as promise the MicroProfile project information for the given
 	 *         application.properties URI.
 	 */
-	public CompletableFuture<MicroProfileProjectInfo> getMicroProfileProjectInfo(MicroProfileProjectInfoParams params) {
-		// Search project info in cache
-		MicroProfileProjectInfoWrapper projectInfo = cache.get(params.getUri());
-		if (projectInfo == null) {
-			// not found in cache, load the project info from the JDT LS Extension
+	public CompletableFuture<MicroProfileProjectInfo> getProjectInfo(MicroProfileProjectInfoParams params) {
+		// Search future which load project info in cache
+		CompletableFuture<MicroProfileProjectInfo> projectInfo = cache.get(params.getUri());
+		if (projectInfo == null || projectInfo.isCancelled() || projectInfo.isCompletedExceptionally()) {
+			// not found in the cache, load the project info from the JDT LS Extension
 			params.setScopes(MicroProfilePropertiesScope.SOURCES_AND_DEPENDENCIES);
-			return provider.getProjectInfo(params).thenApply(info ->
-			// information was loaded, update the cache
-			{
-				cache.put(params.getUri(), new MicroProfileProjectInfoWrapper(info));
-				return info;
-			});
+			CompletableFuture<MicroProfileProjectInfo> future = provider.getProjectInfo(params). //
+					thenApply(info -> new MicroProfileProjectInfoWrapper(info));
+			// cache the future.
+			cache.put(params.getUri(), future);

```

Comment:
where is the entry removed? looks like a memory leak waiting to happen

Reply:
The remove is done when classpath changed. I think the remove should be done when project is closed. Do you think we should manage that?

Reply:
> The remove is done when classpath changed. I think the remove should be done when project is closed. Do you think we should manage that?

yes

Reply:
also, if document is deleted

Reply:
To clarify your comment you wish i remove the cache when application.properties is removed and when project is closed. Is that? If it that i need to manage that on jdt ls extension side because diddelete doesnt exist in lsp

Reply:
@fbricon I create an issue to manage this cache with project events at https://github.com/redhat-developer/quarkus-ls/issues/184